baseplate/lib/secrets.py

```diff
@@ -464,6 +467,91 @@ def _get_file_data(self, filename: str) -> Tuple[Dict, float]:
+
+class VaultCSISecretsStore(SecretsStore):
+    """Access to secret tokens using a vault CSI mount with automatic refresh.
+
+    This store allows access to secrets stored in vault through the CSI interface.
+    It performs caching and will automatically reload secrets from disk. Generally
+    do not cache or store the values returned by this class's methods, rather get
+    them from this class each time you need them. The secrets are served from memory
+    when possible, so there's little performance impact in doing so, and you will be
+    sure to always have the current version in the face of key rotation etc.
+    """
+
+    path: Path
+    data_symlink: Path
+    cache: Dict[str, VaultCSIEntry]
+
+    def __init__(
+        self,
+        path: str,
+        parser: SecretParser,
+    ):  # pylint: disable=super-init-not-called
+        self.path = Path(path)
+        self.parser = parser
+        self.cache = {}
+        self.data_symlink = self.path.joinpath("..data")
+        if not self.path.is_dir():
+            raise ValueError(f"Expected {self.path} to be a directory.")
+        if not self.data_symlink.is_dir():
+            raise ValueError(
+                f"Expected {self.data_symlink} to be a directory. Verify {self.path} is the root of the Vault CSI mount."
+            )
+
+    def get_vault_url(self) -> str:
+        """Deprecated and will be removed in 3.0.0"""
+        raise NotImplementedError
+
+    def get_vault_token(self) -> str:
+        """Deprecated and will be removed in 3.0.0"""
+        raise NotImplementedError
+
+    def _get_mtime(self) -> float:
+        """Modification time is store-wide for CSI secrets."""
+        return os.path.getmtime(self.data_symlink)
+
+    def _raw_secret(self, name: str) -> Any:
+        try:
+            with open(self.data_symlink.joinpath(name), "r", encoding="UTF-8") as fp:
+                return self.parser(json.load(fp))
+        except FileNotFoundError:
+            # Try a second time in case the file was deleted while we tried to read it
+            try:
+                with open(self.data_symlink.joinpath(name), "r", encoding="UTF-8") as fp:
+                    return self.parser(json.load(fp))
+            except FileNotFoundError as exc:
+                raise SecretNotFoundError(name) from exc
+
+    def get_raw_and_mtime(self, secret_path: str) -> Tuple[Dict[str, str], float]:
+        mtime = self._get_mtime()
+        if cache_entry := self.cache.get(secret_path):
+            if cache_entry.mtime == mtime or cache_entry.updating is True:
+                return cache_entry.data, mtime
+            # Prevent multiple requests from updating the cache at the same time by setting the
+            # `updating` flag. Unfortunately we can only prevent this when the cache currently
+            # has a valid entry.
+            self.cache[secret_path] = VaultCSIEntry(
+                mtime=cache_entry.mtime, data=cache_entry.data, updating=True
+            )
```

Comment:
The only real reason in my mind is to prevent multiple disk hits. Since the cache will expire every 2 minutes or so, I was concerned that high traffic services might see relatively frequent IO / latency spikes. I'd also argue that the absolute complexity increase is not very great–a single extra line of code and an extra boolean in the cache entry tuple.

Does any of that change your mind?

Reply:
The problem is the mutex.  Using a mutex _always_ increases complexity beyond what it appears on the page.

Now and into the future, we now have to be triply careful that we always unlock the mutex when we lock it.  Especially with exceptions, that can turn out to be incredibly hard to ensure.

I think I'm actually talking myself _more_ into wanting to remove this, rather than less 😅

Reply:
Removed!